mypy/semanal.py

```diff
@@ -422,25 +422,82 @@ def is_defined_type_var(self, tvar: str, context: Context) -> bool:
         return self.lookup_qualified(tvar, context).kind == BOUND_TVAR
 
     def visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
+        # OverloadedFuncDef refers to any legitimate situation where you have
+        # more than one declaration for the same function in a row.  This occurs
+        # with a @property with a setter or a deleter, and for a classic
+        # @overload.
+
+        # Decide whether to analyze this as a property or an overload.  If an
+        # overload, and we're outside a stub, find the impl and set it.  Remove
+        # the impl from the item list, it's special.
         t = []  # type: List[CallableType]
+        non_overload_indexes = []
         for i, item in enumerate(defn.items):
             # TODO support decorated overloaded functions properly
             item.is_overload = True
-            item.func.is_overload = True
             item.accept(self)
-            callable = function_type(item.func, self.builtin_type('builtins.function'))
-            assert isinstance(callable, CallableType)
-            t.append(callable)
-            if item.func.is_property and i == 0:
-                # This defines a property, probably with a setter and/or deleter.
-                self.analyze_property_with_multi_part_definition(defn)
-                break
-            if not [dec for dec in item.decorators
-                    if refers_to_fullname(dec, 'typing.overload')]:
-                self.fail("'overload' decorator expected", item)
 
-        defn.type = Overloaded(t)
-        defn.type.line = defn.line
+            if isinstance(item, Decorator):
+                item.func.is_overload = True
+                callable = function_type(item.func, self.builtin_type('builtins.function'))
+                assert isinstance(callable, CallableType)
+                if item.func.is_property and i == 0:
+                    # This defines a property, probably with a setter and/or
```

Comment:
Doesn't it make more sense to check for this separately before looping over defn.items? It really seems to be quite a separate case.